src/webgl/ShaderGenerator.js

```diff
@@ -1638,3 +1638,488 @@ export default shadergenerator;
+ * - <a href="#/p5/baseStrokeShader">baseStrokeShader()</a>
+ *
+ * @param {function} callback
+ *        A callback function which is called after each vertex is processed.
+ */
+
+/**
+ * @method beforeFragment
+ * @private
+ * @description
+ * Registers a callback to run custom code at the very start of the fragment shader. This hook can be used inside <a href="#/p5/baseColorShader">baseColorShader()</a>.modify() and similar shader modify calls to set up variables or perform calculations that affect every pixel before color calculations begin. The callback receives no arguments.
+ *
+ * This hook is available in:
+ * - <a href="#/p5/baseColorShader">baseColorShader()</a>
+ * - <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>
+ * - <a href="#/p5/baseNormalShader">baseNormalShader()</a>
+ * - <a href="#/p5/baseStrokeShader">baseStrokeShader()</a>
+ *
+ * @param {function} callback
+ *        A callback function which is called before each fragment is processed.
+ *
+ * @example
+ * <div modernizr='webgl'>
+ * <code>
+ * let myShader;
+ * function setup() {
+ *   createCanvas(200, 200, WEBGL);
+ *   myShader = baseColorShader().modify(() => {
+ *     beforeFragment(() => {
+ *       // Set a value for use in getFinalColor
+ *       this.brightness = 0.5 + 0.5 * sin(millis() * 0.001);
+ *     });
+ *     getFinalColor(color => {
+ *       // Use the value set in beforeFragment to tint the color
+ *       color.r *= this.brightness; // Tint red channel
+ *       return color;
+ *     });
+ *   });
+ * }
+ * function draw() {
+ *   background(220);
+ *   shader(myShader);
+ *   noStroke();
+ *   fill('teal');
+ *   box(100);
+ * }
+ * </code>
+ * </div>
+ */
+
+/**
+ * @method getPixelInputs
+ * @description
+ * Registers a callback to modify the properties of each fragment (pixel) before the final color is calculated in the fragment shader. This hook can be used inside <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>.modify() and similar shader modify calls to change opacity or other per-pixel properties. The callback receives an object with the following properties:
+ * - `opacity`: a number between 0 and 1 for the pixel's transparency
+ * - (Other properties may be available depending on the shader.)
+ *
+ * Return the modified object to update the fragment.
+ *
+ * This hook is available in:
+ * - <a href="#/p5/baseMaterialShader">baseMaterialShader()</a>
+ * - <a href="#/p5/baseStrokeShader">baseStrokeShader()</a>
+ *
+ * @param {function} callback
+ *        A callback function which receives the fragment inputs object and should return it after making any changes.
+ *
+ * @example
+ * <div modernizr='webgl'>
+ * <code>
+ * let myShader;
+ * function setup() {
+ *   createCanvas(200, 200, WEBGL);
+ *   myShader = baseMaterialShader().modify(() => {
+ *     getPixelInputs(inputs => {
+ *       // Animate opacity based on x position
+ *       let t = uniformFloat(() => millis());
```

Comment:
I think the uniform may need to be declared inside of `modify` but outside of `getPixelInputs`.